python/ray/util/sgd/v2/trainer.py

```diff
@@ -226,8 +234,87 @@ def to_tune_trainable(
         Returns:
             A Trainable that can directly be passed into ``tune.run()``.
         """
+        return _create_tune_trainable(train_func, self._backend,
+                                      self._num_workers, self._use_gpu,
+                                      self._resources_per_worker)
+
 
-        def trainable_func(config: Dict[str, Any]) -> T:
-            pass
+class _TuneTrainer(Trainer):
+    def start_training(
+            self,
+            train_func: Union[Callable[[], T], Callable[[Dict[str, Any]], T]],
+            config: Optional[Dict[str, Any]] = None) -> None:
+        train_func = self._get_train_func(train_func, config)
+        self._executor.start_training(train_func)
+
+    def fetch_next_report(self) -> Optional[List[Dict]]:
+        return self._executor.fetch_next_result()
+
+
+class _SgdTrainable(DistributedTrainable):
+    _function = None
+    _backend = None
+    _num_workers = None
+    _use_gpu = None
+    _resources_per_worker = None
+
+    _started = False
+    _finished = False
+
+    def setup(self, config: Dict):
+        # Create a new Trainer with the same configuration.
+        self._trainer = _TuneTrainer(self._backend, self._num_workers,
+                                     self._use_gpu, self._resources_per_worker)
+        # TODO(matt): use placement group resource allocation
+        self._trainer.start()  # Move this to step?
+        self._config = config
+
+    def step(self) -> Dict:
+        if self._finished:
+            raise RuntimeError("Training has already finished.")
+        if not self._started:
+            self._trainer.start_training(self._function, self._config)
+            self._started = True
+
+        report = self._trainer.fetch_next_report()
+        if report is None:
+            self._finished = True
+            return {RESULT_DUPLICATE: True}
+        # Currently return the value from first worker.
+        # TODO(matt): add support for aggregation function.
+        return report[0]
+
+    def save_checkpoint(self, tmp_checkpoint_dir):
+        raise NotImplementedError
 
+    def load_checkpoint(self, checkpoint):
         raise NotImplementedError
+
+    def cleanup(self):
+        self._trainer.shutdown()
+
+
+def _create_tune_trainable(train_func, backend, num_workers, use_gpu,
```

Comment:
Just wondering, what was the decision making for moving this out of the `Trainer`?

Reply:
I ran into a problem when trying to define the `@classmethod` `default_resource_request` using instance variables  from the outer `Trainer` class (e.g. `train_func`). Do you know if there's a cleaner way to do it? I can also update the method doc to make this logic more clear.